src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs

```diff
@@ -0,0 +1,138 @@
+            var members = root.Members;
+            if (members.Any(m => m is GlobalStatementSyntax))
+            {
+                var compilation = await document.Project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+                // While the analyze and refactoring check to ensure we're in a well formed state for their needs, the
+                // 'new template' code just calls directly into this if the user prefers Program.Main.  So check and
+                // make sure this is actually something we can convert before proceeding.
+                var programType = compilation.GetBestTypeByMetadataName(WellKnownMemberNames.TopLevelStatementsEntryPointTypeName);
+                if (programType != null)
+                {
+                    if (programType.GetMembers(WellKnownMemberNames.TopLevelStatementsEntryPointMethodName).FirstOrDefault() is IMethodSymbol mainMethod)
+                    {
+                        var classDeclaration = await GenerateProgramClassAsync(
+                            document, programType, mainMethod, cancellationToken).ConfigureAwait(false);
+
+                        var newRoot = root.RemoveNodes(members.OfType<GlobalStatementSyntax>().Skip(1), SyntaxGenerator.DefaultRemoveOptions);
+                        Contract.ThrowIfNull(newRoot);
+
+                        var firstGlobalStatement = newRoot.Members.OfType<GlobalStatementSyntax>().Single();
+                        newRoot = newRoot.ReplaceNode(
+                            firstGlobalStatement,
+                            FixupComments(classDeclaration.WithLeadingTrivia(firstGlobalStatement.GetLeadingTrivia())));
+
+                        return document.WithSyntaxRoot(newRoot);
+                    }
+                }
+            }
+
+            return document;
+        }
+
+        private static async Task<ClassDeclarationSyntax> GenerateProgramClassAsync(
+            Document document,
+            INamedTypeSymbol programType,
+            IMethodSymbol mainMethod,
+            CancellationToken cancellationToken)
+        {
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+            // Respect user settings on if they want explicit or implicit accessibility modifiers.
+            var option = options.GetOption(CodeStyleOptions2.RequireAccessibilityModifiers);
+            var accessibilityModifiersRequired = option.Value is AccessibilityModifiersRequired.ForNonInterfaceMembers or AccessibilityModifiersRequired.Always;
+
+            var root = (CompilationUnitSyntax)await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var generator = document.GetRequiredLanguageService<SyntaxGenerator>();
+
+            // See if we have an existing part in another file.  If so, we'll have to generate our declaration as partial.
```

Comment:
not sure what you're referring to.  we don't do any DeclaringSyntaxReference checks for converting to Program.Main.  Indeed, there are practically no blocks on that since we can practically always convert top level statemnts to Program.Main :)

(the reverse is not true.  we have a lot of restrictions on converting the opposite direction).

Reply:
Yep, you're right. Went cross-eyed and forgot which direction this was going.